crates/uv/src/commands/tool/upgrade.rs

```diff
@@ -170,32 +245,85 @@ async fn upgrade_tool(
     // Initialize any shared state.
     let state = SharedState::default();
 
-    // TODO(zanieb): Build the environment in the cache directory then copy into the tool
-    // directory.
-    let EnvironmentUpdate {
-        environment,
-        changelog,
-    } = update_environment(
-        existing_environment,
-        spec,
-        &settings,
-        &state,
-        Box::new(SummaryResolveLogger),
-        Box::new(UpgradeInstallLogger::new(name.clone())),
-        connectivity,
-        concurrency,
-        native_tls,
-        cache,
-        printer,
-    )
-    .await?;
-
-    // If we modified the target tool, reinstall the entrypoints.
-    if changelog.includes(name) {
+    // Check if we need to create a new environment — if so, resolve it first, then
+    // install the requested tool
+    let (environment, outcome) = if let Some(interpreter) =
+        interpreter.filter(|interpreter| !environment.uses(interpreter))
+    {
+        // If we're using a new interpreter, re-create the environment for each tool.
+        let resolution = resolve_environment(
+            RequirementsSpecification::from_requirements(requirements.to_vec()).into(),
+            interpreter,
+            settings.as_ref().into(),
+            &state,
+            Box::new(SummaryResolveLogger),
+            connectivity,
+            concurrency,
+            native_tls,
+            cache,
+            printer,
+        )
+        .await?;
+
+        let environment = installed_tools.create_environment(name, interpreter.clone())?;
+
+        let environment = sync_environment(
+            environment,
+            &resolution.into(),
+            settings.as_ref().into(),
+            &state,
+            Box::new(DefaultInstallLogger),
+            connectivity,
+            concurrency,
+            native_tls,
+            cache,
+            printer,
+        )
+        .await?;
+
+        (environment, UpgradeOutcome::UpgradeEnvironment)
+    } else {
+        // Otherwise, upgrade the existing environment.
+        // TODO(zanieb): Build the environment in the cache directory then copy into the tool
+        // directory.
+        let EnvironmentUpdate {
+            environment,
+            changelog,
+        } = update_environment(
+            environment,
+            spec,
+            &settings,
+            &state,
+            Box::new(SummaryResolveLogger),
+            Box::new(UpgradeInstallLogger::new(name.clone())),
+            connectivity,
+            concurrency,
+            native_tls,
+            cache,
+            printer,
+        )
+        .await?;
+
+        let outcome = if changelog.includes(name) {
+            UpgradeOutcome::UpgradeTool
+        } else if changelog.is_empty() {
+            UpgradeOutcome::NoOp
+        } else {
+            UpgradeOutcome::UpgradeDependencies
+        };
+
+        (environment, outcome)
+    };
```

Comment:
I generally tried to structure this such that we avoid all `mut`, and have two branches that return the values we care about.